modules/internal/DefaultAssociative.chpl

```diff
@@ -566,14 +566,15 @@ module DefaultAssociative {
 
     // ref version
     proc dsiAccess(idx : idxType, haveLock = false) ref {
-      const shouldLock = dom.parSafe && !haveLock;
+      const arrOwnsDom = dom._arrs.length == 1;
+      const shouldLock = dom.parSafe && !haveLock && arrOwnsDom;
```

Comment:
>  I don't think we can tell whether an array's domain is const or not from within module code.

not yet, but it's been coming up a lot lately....

Anyway, regarding the rest... it seems like a slippery slope.
What if I wrote

``` chapel
var total: atomic int;
forall key in collection {
  if A.domain.member(key) {
    total += A[key];
  } else {
    A[key] = 0;
  }
}
```

? Wouldn't that be really confusing too? I'm trying to construct an example where there is a race between unlocked reads and locked writes. (Something I was thinking about before... being able to say an operation is "atomic on" some data structure, would help here. You'd just make the whole loop body "atomic on" the domain).

Maybe the answer is to just say that the 1:1 associative array key creation is not parallel safe.

We've talked before about how it'd probably be better to make such array-operations-modifying-domains just be their own separate array types.